File: NumberTheory.hs
-- | NumberTheory.hs
--
-- Number-theory utilities for the Sylow solver.
-- This module provides efficient helpers for primes, divisors and
-- prime factorizations used by the proof engine.
--
-- This is a Haskell port of the Python sylow2.py module.

module NumberTheory
  ( isPrime
  , primes
  , divisors
  , maxPDivisor
  , primeFactors
  , primeFactorization
  , prime
  , numSylow
  , pKillable
  , sylowKillable
  ) where

import Data.List (sort, nub)

-- | Return True if n is prime.
-- Uses simple deterministic trial division (sufficient for moderate n).
isPrime :: Int -> Bool
isPrime n
  | n <= 1 = False
  | n <= 3 = True
  | n `mod` 2 == 0 = False
  | otherwise = not $ any (\i -> n `mod` i == 0) [3, 5 .. isqrt n]
  where
    isqrt = floor . sqrt . fromIntegral

-- | Return list of primes up to and including n using a simple sieve.
primes :: Int -> [Int]
primes n
  | n < 2 = []
  | otherwise = sieve [2..n]
  where
    sieve [] = []
    sieve (p:xs) = p : sieve [x | x <- xs, x `mod` p /= 0]

-- | Return the divisors of n in sorted order.
divisors :: Int -> [Int]
divisors n
  | n <= 0 = []
  | otherwise = sort $ concatMap getDivisorPair [1 .. isqrt n]
  where
    isqrt = floor . sqrt . fromIntegral
    getDivisorPair i
      | n `mod` i == 0 = 
          let j = n `div` i
          in if i == j then [i] else [i, j]
      | otherwise = []

-- | Return k maximal such that p^k divides n.
maxPDivisor :: Int -> Int -> Int
maxPDivisor n p = go n 0
  where
    go m k
      | m `mod` p == 0 = go (m `div` p) (k + 1)
      | otherwise = k

-- | Return the list of prime divisors of n in ascending order.
-- This is more efficient than testing every integer up to n.
primeFactors :: Int -> [Int]
primeFactors n
  | n <= 1 = []
  | otherwise = sort $ factor2 n []
  where
    factor2 m acc
      | m `mod` 2 == 0 = factor2 (removeFactors m 2) (2 : acc)
      | otherwise = factorOdd m 3 acc
    
    factorOdd m p acc
      | p * p > m = if m > 1 then m : acc else acc
      | m `mod` p == 0 = factorOdd (removeFactors m p) (p + 2) (p : acc)
      | otherwise = factorOdd m (p + 2) acc
    
    removeFactors m p
      | m `mod` p == 0 = removeFactors (m `div` p) p
      | otherwise = m

-- | Return prime factorization as a list of (prime, exponent).
-- Example: 12 -> [(2,2),(3,1)].
primeFactorization :: Int -> [(Int, Int)]
primeFactorization n = map (\p -> (p, maxPDivisor n p)) (nub $ primeFactors n)

-- | Compatibility wrapper for isPrime.
prime :: Int -> Bool
prime = isPrime

-- | Return divisors d of n with d ≡ 1 (mod p).
numSylow :: Int -> Int -> [Int]
numSylow p n = filter (\d -> d `mod` p == 1) (divisors n)

-- | Return True if there is no divisor d>1 of n with d ≡ 1 (mod p).
-- Check all divisors except 1.
pKillable :: Int -> Int -> Bool
pKillable p n = not $ any (\d -> d > 1 && d `mod` p == 1) (divisors n)

-- | Return True if there exists a prime p dividing n for which pKillable p n is True.
sylowKillable :: Int -> Bool
sylowKillable n
  | n == 1 = True
  | otherwise = any (`pKillable` n) (reverse $ primeFactors n)

File: sylow_solver/number_theory.py
"""Number-theoretic utilities for Sylow theorem computations."""

from typing import List, Tuple


def divisors(n: int) -> List[int]:
    """Computes the sorted list of divisors of n in O(sqrt(n)) time."""
    if n <= 0:
        return []
    divs = []
    i = 1
    while i * i <= n:
        if n % i == 0:
            divs.append(i)
            if i != n // i:
                divs.append(n // i)
        i += 1
    return sorted(divs)


def is_prime(p: int) -> bool:
    """Determines if a number is prime using trial division."""
    if p < 2:
        return False
    if p == 2:
        return True
    if p % 2 == 0:
        return False
    i = 3
    while i * i <= p:
        if p % i == 0:
            return False
        i += 2
    return True


def primes(n: int) -> List[int]:
    """Generates a list of prime numbers up to and including n."""
    if n < 2:
        return []
    sieve = [True] * (n + 1)
    sieve[0] = sieve[1] = False
    i = 2
    while i * i <= n:
        if sieve[i]:
            for j in range(i * i, n + 1, i):
                sieve[j] = False
        i += 1
    return [i for i, is_p in enumerate(sieve) if is_p]


def max_p_divisor(n: int, p: int) -> int:
    """Finds the maximum k such that p^k divides n."""
    k = 0
    while n % p == 0:
        k += 1
        n //= p
    return k


def prime_factors(n: int) -> List[int]:
    """Computes the sorted list of prime factors of n in O(sqrt(n)) time."""
    if n <= 1:
        return []
    factors = []
    if n % 2 == 0:
        factors.append(2)
        while n % 2 == 0:
            n //= 2
    i = 3
    while i * i <= n:
        if n % i == 0:
            factors.append(i)
            while n % i == 0:
                n //= i
        i += 2
    if n > 1:
        factors.append(n)
    return factors


def prime_factorization(n: int) -> List[Tuple[int, int]]:
    """Computes the prime factorization of n as a list of (prime, exponent) tuples."""
    if n <= 1:
        return []
    factorization = []
    if n % 2 == 0:
        exp = 0
        while n % 2 == 0:
            exp += 1
            n //= 2
        factorization.append((2, exp))
    i = 3
    while i * i <= n:
        if n % i == 0:
            exp = 0
            while n % i == 0:
                exp += 1
                n //= i
            factorization.append((i, exp))
        i += 2
    if n > 1:
        factorization.append((n, 1))
    return factorization


def num_sylow(p: int, n: int) -> List[int]:
    """Computes possible numbers of Sylow p-subgroups in a group of order n."""
    return [d for d in divisors(n) if d % p == 1]


def p_killable(p: int, n: int) -> bool:
    """Determines if the Sylow p-subgroup is automatically normal."""
    for d in divisors(n):
        if d > 1 and d % p == 1:
            return False
    return True


def sylow_killable(n: int) -> bool:
    """Determines if some Sylow p-subgroup is automatically normal."""
    if n == 1:
        return True
    p_factors = prime_factors(n)
    p_factors.reverse()
    for p in p_factors:
        if p_killable(p, n):
            return True
    return False
